pkg/capacity/capacity.go

```diff
@@ -0,0 +1,627 @@
+	go c.topologyInformer.Run(ctx)
+
+	c.prepare(ctx)
+	for i := 0; i < threadiness; i++ {
+		go wait.UntilWithContext(ctx, func(ctx context.Context) {
+			c.runWorker(ctx)
+		}, time.Second)
+	}
+
+	go wait.UntilWithContext(ctx, func(ctx context.Context) { c.pollCapacities() }, c.pollPeriod)
+
+	klog.Info("Started Capacity Controller")
+	<-ctx.Done()
+	klog.Info("Shutting down Capacity Controller")
+}
+
+func (c *Controller) prepare(ctx context.Context) {
+	// Wait for topology and storage class informer sync. Once we have that,
+	// we know which CSIStorageCapacity objects we need.
+	if !cache.WaitForCacheSync(ctx.Done(), c.topologyInformer.HasSynced, c.scInformer.Informer().HasSynced, c.cInformer.Informer().HasSynced) {
+		return
+	}
+
+	// The caches are fully populated now, but the event handlers
+	// may or may not have been invoked yet. To be sure that we
+	// have all data, we need to list all resources. Here we list
+	// topology segments, onTopologyChanges lists the classes.
+	c.onTopologyChanges(c.topologyInformer.List(), nil)
+
+	klog.V(3).Infof("Initial number of topology segments %d, storage classes %d, potential CSIStorageCapacity objects %d",
+		len(c.topologyInformer.List()),
+		func() int {
+			scs, _ := c.scInformer.Lister().List(labels.Everything())
+			return len(scs)
+		}(),
+		len(c.capacities))
+
+	// Now that we know what we need, we can check what we have.
+	// We do that both via callbacks *and* by iterating over all
+	// objects: callbacks handle future updates and iterating
+	// avoids the assumumption that the callback will be invoked
+	// for all objects immediately when adding it.
+	klog.V(3).Info("Checking for existing CSIStorageCapacity objects")
+	handler := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			csc, ok := obj.(*storagev1alpha1.CSIStorageCapacity)
+			if !ok {
+				klog.Errorf("added object: expected CSIStorageCapacity, got %T -> ignoring it", obj)
+				return
+			}
+			c.onCAddOrUpdate(ctx, csc)
+		},
+		UpdateFunc: func(_ interface{}, newObj interface{}) {
+			csc, ok := newObj.(*storagev1alpha1.CSIStorageCapacity)
+			if !ok {
+				klog.Errorf("updated object: expected CSIStorageCapacity, got %T -> ignoring it", newObj)
+				return
+			}
+			c.onCAddOrUpdate(ctx, csc)
+		},
+		DeleteFunc: func(obj interface{}) {
+			// Beware of "xxx deleted" events
+			if unknown, ok := obj.(cache.DeletedFinalStateUnknown); ok && unknown.Obj != nil {
+				obj = unknown.Obj
+			}
+			csc, ok := obj.(*storagev1alpha1.CSIStorageCapacity)
+			if !ok {
+				klog.Errorf("deleted object: expected CSIStorageCapacity, got %T -> ignoring it", obj)
+				return
+			}
+			c.onCDelete(ctx, csc)
+		},
+	}
+	c.cInformer.Informer().AddEventHandler(handler)
+	capacities, err := c.cInformer.Lister().List(labels.Everything())
+	if err != nil {
+		// This shouldn't happen.
+		utilruntime.HandleError(err)
+		return
+	}
+	for _, capacity := range capacities {
+		c.onCAddOrUpdate(ctx, capacity)
```

Comment:
Still outstanding. It would be nice to have an explanation why it's important we invoke all the event handlers initially.

Reply:
Sorry, missed this one. There was already a comment further up (" Now that we know what we need, we can check what we have.") which I extended:

> This is important to avoid duplicates, i.e. creating a new
> object with the same topology and storage class as an
> existing one instead of updating that one.

The reason for explicitly iterating over the list was also explained there ("We do that both via callbacks *and* by iterating over all objects: callbacks handle future updates and iterating avoids the assumption that the callback will be invoked for all objects immediately when adding it.").

Without iterating over the list I had seen random test failures because the event handlers where getting called in parallel to and sometimes after the initial processing. This is consistent with the documentation of informers, which only states that the local state inside the cache is up-to-date after WaitForCacheSync.